### DISCUSSION TITLE
How does Better Bibtex deal with pinned citation keys synced from mutiple computers?

### DISCUSSION THREAD

**Discussion by earthmoon** [retorquere/zotero-better-bibtex#2368] [Q&A]
> Depends on the actual scenario. If you have auto-pin on, this will happen:
> 
> Computer A: creates ABC123
> Computer A: pins citekey to "1" (for example)
> Computer A: syncs
> Computer B: syncs
> Computer B: received ABC123 with pinned key "1"
> 
> If you don't have auto-pin on, you could have this scenario

> Computer A: creates ABC123 with a non-pinned citekey
> Computer A: syncs
> Computer B: syncs
> Computer B: receives ABC123 and creates a new non-pinned citekey
> Computer A: you pin the key to "1"
> Computer B: you ping the key to "2"
> Computer A: syncs
> Computer B: syncs
> 
> In the last step you should see a sync conflict, as the item was changed on both sides. You get to choose which fields to keep in the merge. Whichever extra field you choose to keep, that citekey will be used.

I am a little confused:

> Computer A: creates ABC123
> Computer A: pins citekey to "1" (for example)
> Computer A: syncs
> Computer B: syncs
> Computer B: received ABC123 with pinned key "1"

In this case, if pin keys to item(id) in Zotero, on computers A and B, they wouldn't have the same keys, as the item(id) in computer A and computer B is different. If they wouldn't have the same keys, how Better Bibtex deal with files synced? For example:
>Computer A: creates **123** for the reference **A**, **123** is the item(id)
>Computer A: pins citation key to **123**
>Computer A: syncs
>Computer B: syncs
>Computer B:  the item(id) of the reference **A** wouldn't be **123**, as item(id) wouldn't be synced in Zotero. Here, let **456** be the item(id) on computer B.

1. Computer B: receive reference **A**. According to the pin rules, it should use item(id) on computer B to create reference **A** a citation key, So, the reference **A** has a citation key of **456**; Am I right?
2. Computer B: receive reference **A**.  As Better Bibtex files had been synced to computer B, so Better Bibtex would check the files synced, and find that the citation key of the reference **A** is pinned! So, the citation keys pinned shouldn't be changed. Better Bibtex should keep reference **A** citation key to **123**.

You see the first situation differentiates from the second situation. There are no violations of rules in both case, and they have all violated the rules. In the first situation, Better Bibtex uses item(id), but how to deal with citation keys synced? In the second situation, Better Bibtex uses files synced, which is the value of synchronization, but it doesn't use item(id), which is weird as it had been set up to use **id** but actually does not use **id**.

> Computer A: creates ABC123 with a non-pinned citekey
> Computer A: syncs
> Computer B: syncs
> Computer B: receives ABC123 and creates a new non-pinned citekey
> Computer A: you pin the key to "1"
> Computer B: you ping the key to "2"
> Computer A: syncs
> Computer B: syncs
> 
> In the last step you should see a sync conflict, as the item was changed on both sides. You get to choose which fields to keep in the merge. Whichever extra field you choose to keep, that citekey will be used.

In this situation, conflicts can be accepted. 

May you share with me what reference manager are you using now, and why do you choose it?

**Comment by retorquere** [comment]
> Computer A: creates 123 for the reference A, 123 is the item(id)
> Computer A: pins citation key to 123
> Computer A: syncs
> Computer B: syncs
> Computer B: the item(id) of the reference A wouldn't be 123, as item(id) wouldn't be synced in Zotero. Here, let 456 be the item(id) on computer B.
> 
>  Computer B: receive reference A. According to the pin rules, it should use item(id) on computer B to create reference A a citation key, So, the reference A has a citation key of 456; Am I right?

No. Here is what actually happens.

* Computer A: creates item { id: 123, key: 'ABCD', extra: ''}, nothing in extra, so gets dynamic key `123`
* Computer A: pins citation key to 123, item is now { id: 123, key: 'ABCD', extra: 'Citation Key: 123'}, the presence of the 'Citation Key' in extra means that key is pinned to `123`
* Computer A: syncs item { key: 'ABCD', extra: 'Citation Key: 123'}
* Computer B: syncs and receives { key: 'ABCD', extra: 'Citation Key: 123'}, resulting in { id: 456, key: 'ABCD', extra: 'Citation Key: 123'}, **the presence of the 'Citation Key' in extra means that key is pinned to `123`**

> May you share with me what reference manager are you using now, and why do you choose it?

Only Zotero. No reference manager is a perfect fit for users' specific use-cases, but Zotero allows the community to address those functionality gaps with a crazy extensive plugin facility. Not one other ref manager allows this. Of the rest, Mendeley gets special mention for their [user-hostile](https://twitter.com/search?q=mendeley%20gdpr&src=typed_query&f=top) practices, where they lock people out of their own data, "because GDPR", which they have refused to clarify 4 years and counting.

**Reply by earthmoon** [reply]
> Computer B: syncs and receives { key: 'ABCD', extra: 'Citation Key: 123'}, resulting in { id: 456, key: 'ABCD', extra: 'Citation Key: 123'}, **the presence of the 'Citation Key' in extra means that key is pinned to** `123`

So, once a citation key is pinned to item(id) to a reference, the citation key will be unique even on another computer. But if the citation key on computer B is refreshed, will it be changed to `456` as item(id) in computer B?
From the perspective of pinned citation keys, item(key) is better than item(id), as item(key) will be synced and wouldn't be different on different computers. However, item(key)'s recognition is not high as item(id), as item(id) can be remembered easily even not need to copy it but see it and input it in the Latex file. So, why did you once suggest I use item(id) not use item(key). Confused again.

**Reply by retorquere** [reply]
> So, once a citation key is pinned to item(id) to a reference, the citation key will be unique even on another computer.

Not specific to `item(id)`, but yes, barring sync conflicts as described earlier. If you do the pinning on more than one computer, it is in principle possible that you'll get keys like `123a`. If you pin on one computer, that won't occur.

> But if the citation key on computer B is refreshed, will it be changed to 456 as item(id) in computer B?

No, pinned citekeys are not refreshed. If you want to refresh, you should unpin, and then refresh (and optionally pin again). If you do that, it will indeed be pinned to `456`.

**Reply by retorquere** [reply]
> From the perspective of pinned citation keys, item(key) is better than item(id), as item(key) will be synced and wouldn't be different on different computers. However, item(key)'s recognition is not high as item(id), as item(id) can be remembered easily even not need to copy it but see it and input it in the Latex file. So, why did you once suggest I use item(id) not use item(key). Confused again.

I did initially say that you should use `item` (which is shorthand for `item(key)`). You later stated that you preferred numeric keys, and **if** you want numeric keys, **then** you should use `item(id)`.

**Reply by earthmoon** [reply]
The references are pinned by item(id). If the user accidentally clicks the wrong button, for example click unpin for one or more references. The references will be refreshed by a new pinning rule set in the preference of the Better Bibtex. When the user finds this phenomenon, he/her wants to restore the citation keys to item(id). He/her may not be able to do that, as in which computer the references were pinned by item(id) may not be remembered and confirmed. So, pinning references with item(id) is not safe as with item(key)

However, item(id) is much more readable than item(key). As for me, safety is more important than readability, so after having learnt so much from your discussion, I still will use item(key) to pin citation keys.

Thanks.

**Reply by retorquere** [reply]
This is all correct. I want to note here that, in the comparison to endnote, where this discussion started, this scenario wouldn't be relevant, because the endnote record number wouldn't work across synced PCs. If you don't sync Zotero, `item(id)` is just as stable as `item(key)`.

**Reply by earthmoon** [reply]
To answer your reviews about Endnote accurately, I have installed Endnote on two Windows 11 computers.

1. Started Endnote on computer A, created a new library, inserted a new reference and saw the record number of the references was 1;
2. Started Endnote on computer B, created a new library, inserted another new reference and saw the record number of the references was 1;
3. Synced the library of Endnote on computer A;
4. Synced the library of Endnote on computer B;

May you guess what happened?  The Endote didn't show synchronization conflict but even didn't prompt any information about this conflict. So, what did Endnote do to solve the two references having the same record number?  It directly changed the record number of one of two references. This behaviour is unacceptable. Choose Zotero not choose Endnote.
